src/mpid/ch4/netmod/ofi/ofi_rma.c

```diff
@@ -118,11 +118,183 @@ int MPIDI_OFI_nopack_putget(const void *origin_addr, int origin_count,
+            MPIDI_OFI_CALL_RETRY(fi_writemsg(MPIDI_OFI_WIN(win).ep, &msg, flags), rdma_write,
+                                 FALSE);
+        else    /* MPIDI_OFI_GET */
+            MPIDI_OFI_CALL_RETRY(fi_readmsg(MPIDI_OFI_WIN(win).ep, &msg, flags), rdma_write, FALSE);
+
+        i += msg_len;
+
+        if (msg_len < req->noncontig.putget.target_iov[target_cur].iov_len) {
+            req->noncontig.putget.target_iov[target_cur].iov_base =
+                (char *) req->noncontig.putget.target_iov[target_cur].iov_base + msg_len;
+            req->noncontig.putget.target_iov[target_cur].iov_len -= msg_len;
+        } else {
+            j++;
+        }
+    }
+    MPIR_Assert(i == actual_pack_bytes);
+    req->noncontig.putget.origin_pack_offset += actual_pack_bytes;
+
+    /* more to send. update our place in the target iov array for later. */
+    if (req->noncontig.putget.origin_pack_offset < req->noncontig.putget.origin_bytes) {
+        req->noncontig.putget.target_iov_cur = j;
+    } else {
+        /* finished issuing. move from deferredQ to syncQ. */
+        DL_DELETE(MPIDI_OFI_WIN(win).deferredQ, req);
+        req->next = MPIDI_OFI_WIN(win).syncQ;
+        MPIDI_OFI_WIN(win).syncQ = req;
+        MPIDI_OFI_sigreq_complete(sigreq);
+        MPL_free(req->noncontig.putget.target_iov);
+    }
+
+  fn_exit:
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPIDI_OFI_pack_putget(const void *origin_addr, int origin_count,
+                          MPI_Datatype origin_datatype, int target_rank,
+                          MPI_Aint target_disp, int target_count,
+                          MPI_Datatype target_datatype, MPIR_Win * win,
+                          MPIDI_av_entry_t * addr, int rma_type, MPIR_Request ** sigreq)
+{
+    int mpi_errno = MPI_SUCCESS;
+    size_t target_bytes, origin_bytes;
+
+    MPIR_Datatype_get_size_macro(origin_datatype, origin_bytes);
+    origin_bytes *= origin_count;
+    MPIR_Datatype_get_size_macro(target_datatype, target_bytes);
+    target_bytes *= target_count;
+
+    /* allocate request */
+    MPIDI_OFI_win_request_t *req = MPIDI_OFI_win_request_create();
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+    req->event_id = MPIDI_OFI_EVENT_ABORT;
+
+    /* allocate target iovecs */
+    struct iovec *target_iov;
+    void *target_base;
+    MPI_Aint total_target_iov_len;
+    MPI_Aint target_len;
+    MPIR_Typerep_iov_len(target_count, target_datatype, target_bytes, &total_target_iov_len);
+    target_len = MPL_MIN(total_target_iov_len, MPIR_CVAR_CH4_OFI_RMA_IOVEC_MAX);
+    target_iov = MPL_malloc(sizeof(struct iovec) * target_len, MPL_MEM_RMA);
+    target_base =
+        (void *) (MPIDI_OFI_winfo_base(win, target_rank) +
+                  +(target_disp * MPIDI_OFI_winfo_disp_unit(win, target_rank)));
+
+    /* allocate pack buffer */
+    MPI_Aint pack_bytes = MPL_MIN(origin_bytes, MPIDI_OFI_global.max_msg_size);
```

Comment:
For the providers that we test with, `MPIDI_OFI_global.max_msg_size` is `INT_MAX`.  So, this would never chop the data into smaller segments.  Instead, how about introduce a macro for the size?  For now, you can set it to some number (say `65536`), but once the genQ functionality is in, it'll be whatever that cell size is (probably also `65536`).

Reply:
Change to `64 * 1024` and added FIXME to use genq API.